File: exam/src/main/java/in/testpress/exam/ui/view/WebView.java
package in.testpress.exam.ui.view;

import android.annotation.TargetApi;
import android.content.Context;
import android.os.Build;
import android.util.AttributeSet;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputConnection;

import static android.text.InputType.TYPE_CLASS_NUMBER;
import static android.text.InputType.TYPE_CLASS_TEXT;
import static android.text.InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD;

public class WebView extends android.webkit.WebView {

    private boolean numberType;

    public WebView(Context context) {
        super(context);
    }

    public WebView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public WebView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public WebView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    public InputConnection onCreateInputConnection(EditorInfo outAttrs) {
        InputConnection inputConnection = super.onCreateInputConnection(outAttrs);
        int type = numberType ? TYPE_CLASS_NUMBER : TYPE_CLASS_TEXT;
        outAttrs.inputType = type | TYPE_TEXT_VARIATION_VISIBLE_PASSWORD;
        return inputConnection;
    }

    public void setNumberType(boolean numberType) {
        this.numberType = numberType;
    }
}

File: core/src/main/java/in/testpress/util/webview/WebView.kt
package `in`.testpress.util.webview

import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.view.View
import android.webkit.CookieManager
import android.webkit.WebSettings
import android.webkit.WebView as AndroidWebView
import `in`.testpress.util.HtmlResourceProcessor
import `in`.testpress.util.LocalWebFileCache

open class WebView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : AndroidWebView(context, attrs, defStyleAttr) {
    
    init {
        configureDefaultSettings()
        configureFocusAndTouch()
    }
    
    private fun configureDefaultSettings() {
        settings.apply {
            javaScriptEnabled = true
            domStorageEnabled = true
            useWideViewPort = false
            loadWithOverviewMode = true
            builtInZoomControls = false
            cacheMode = WebSettings.LOAD_CACHE_ELSE_NETWORK
            setSupportZoom(false)
            userAgentString += " TestpressAndroidApp/WebView"
        }
        
        clearCache(true)
        clearHistory()
        CookieManager.getInstance().setAcceptThirdPartyCookies(this, true)
        
        if (0 != (context.applicationInfo.flags and android.content.pm.ApplicationInfo.FLAG_DEBUGGABLE)) {
            setWebContentsDebuggingEnabled(true)
        }
    }
    
    private fun configureFocusAndTouch() {
        // Input and focus handling (matches XML defaults for interactive views)
        isFocusable = true
        isFocusableInTouchMode = true
        isClickable = true
        isLongClickable = true
        isSaveEnabled = true
        
        // Accessibility (Android O+)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            importantForAutofill = View.IMPORTANT_FOR_AUTOFILL_YES
        }
        importantForAccessibility = View.IMPORTANT_FOR_ACCESSIBILITY_YES
        
        // Request initial focus
        requestFocus()
    }
    
    fun enableFileAccess() {
        settings.apply {
            allowFileAccess = true
            allowFileAccessFromFileURLs = true
            allowUniversalAccessFromFileURLs = true
            mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
        }
    }
    
    fun loadTemplateAndCacheResources(
        templateName: String,
        replacements: Map<String, String>,
        baseUrl: String
    ) {
        val template = LocalWebFileCache.loadTemplate(context, templateName, replacements)
        val resources = HtmlResourceProcessor.extractExternalResources(template)
        
        if (resources.isNotEmpty()) {
            LocalWebFileCache.downloadMultipleInBackground(context, resources)
        }
        
        val localPaths = resources.associate { (url, fileName) ->
            url to LocalWebFileCache.getLocalPath(context, fileName, url)
        }
        
        val processedHtml = HtmlResourceProcessor.replaceUrls(template, localPaths)
        loadDataWithBaseURL(baseUrl, processedHtml, "text/html", "UTF-8", null)
    }
}
